packages/next/src/server/app-render/app-render.tsx

```diff
@@ -1478,11 +1479,13 @@ export async function renderToHTMLOrFlight(
               )
             } else if (isRedirectError(error)) {
               const redirectUrl = getURLFromRedirectError(error)
+              const isPermanent =
+                getRedirectStatusCodeFromError(error) === 308 ? true : false
               if (redirectUrl) {
                 errorMetaTags.push(
                   <meta
                     httpEquiv="refresh"
-                    content={`0;url=${redirectUrl}`}
+                    content={`${isPermanent ? 0 : 1};url=${redirectUrl}`}
```

Comment:
From what I understand setting this number to 1 means the browser will wait 1 second before redirecting, are you aware of that?

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meta

Reply:
This is intentional. Googlebot considers `0` to be permanent and anything above 0 to be temporary: https://developers.google.com/search/docs/crawling-indexing/301-redirects

It's an SEO thing basically, unfortunately Googlebot doesn't support `http-equiv` `status` which would allow us to do this while preserving instant redirect.

Reply:
Hm, not ideal for auth gating ([temporary `redirect()` in either in `page.tsx` or in `layout.tsx`](https://github.com/search?q=repo%3Aupleveled%2Fnext-js-example-spring-2023-vienna-austria+redirect%28&type=code)):

- unusual, slow UX (waiting for 1 second)
- semantically, a permanent redirect is not correct here
- often Middleware is suggested here, but Middleware cannot be used with database auth due to [lack of database TCP connection support](https://github.com/vercel/next.js/discussions/46722)

ideally, either Middleware or a "React Float"-style proper [`Location` response header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Location) would be solutions here

But I guess we'll be recommending `permanentRedirect()` to our students for now to avoid the 1s weirdness

Reply:
A simple solution would be to redirect from JavaScript if there is the refresh meta tag so you don’t need to wait the 1 second delay, not sure if the router already does this

Reply:
The tricky part is around streaming in general. If it's used in a streaming context and HTTP headers are already sent to the client (e.g., with status 200), we can't emit additional headers afterward. This is why meta tag is used this way as far as I understand.

Google warns Javascript redirect to be the last resort to adopt so I guess it's safe to avoid that:
https://developers.google.com/search/docs/crawling-indexing/301-redirects#jslocation

Reply:
> If it's used in a streaming context and HTTP headers are already sent to the client (e.g., with status 200), we can't emit additional headers afterward

Right, with [my suggestion B above](https://github.com/vercel/next.js/pull/54047#discussion_r1309897346) I meant to "lift up" the `redirect()` / `permanentRedirect()` and logic to an earlier part in the response cycle, so that the status code could be set and the `Location` header

Reply:
We use pages / layouts auth gating because Middleware is not an option ([pages / layout auth gating also recommended here as an alternative by Lee](https://github.com/vercel/next.js/discussions/46722#:~:text=It%27s%20also%20worth,set%20of%20routes.))

But [the Edge Runtime Middleware is not capable of doing this with database auth](https://github.com/vercel/next.js/discussions/46722), if you need a TCP connection, which most database options do. So users of Next.js App Router are currently stuck with pages / layout.

However, I would love to be able to do this with Middleware, eg. if there will be tangible movement soon about [this point from Lee](https://github.com/vercel/next.js/discussions/46722#:~:text=exploring%20ways%20to%20bridge%20the%20gap%20towards%20allowing%20folks%20to%20experimentally/%22dangerously%22%20opt%2Dinto%20using%20the%20Node.js%20Runtime):

> exploring ways to bridge the gap towards allowing folks to experimentally/"dangerously" opt-into using the Node.js Runtime